### PR TITLE
Fixes for Charm

### PR DESCRIPTION
**Pull request by ymkc**
- make certain mob families uncharmable to match retail (#2943)
- duration of charm-bind randomized

Used wikia as the main source for the families, left abyssea mostly untouched

### CODE REVIEW

**Comment by teschnei** [comment]
We're not going to stick a giant array of every uncharmable family into the core, sorry.  We can add a flag to one of the sql tables and check that, though.

File: src/map/utils/battleutils.cpp [hunk 1]
```diff
@@ -3900,19 +3901,58 @@ namespace battleutils
...
                 PVictim->m_EcoSystem == SYSTEM_ELEMENTAL || PVictim->m_EcoSystem == SYSTEM_EMPTY ||
                 PVictim->m_EcoSystem == SYSTEM_LUMORIAN || PVictim->m_EcoSystem == SYSTEM_LUMINION ||
-                PVictim->m_EcoSystem == SYSTEM_UNDEAD || PVictim->PMaster != nullptr) {
+                PVictim->m_EcoSystem == SYSTEM_UNDEAD || PVictim->m_EcoSystem == SYSTEM_DRAGON ||
+                PVictim->m_EcoSystem == SYSTEM_HUMANOID ||
+                std::find(std::begin(families), std::end(families), ((CMobEntity*)PVictim)->m_Family) != std::end(families) || 
+                PVictim->PMaster != nullptr) {
                 PVictim->StatusEffectContainer->AddStatusEffect(
-                    new CStatusEffect(EFFECT_BIND, EFFECT_BIND, 1, 0, 5));
+                    new CStatusEffect(EFFECT_BIND, EFFECT_BIND, 1, 0, dsprand::GetRandomNumber(0, 5)));
```
Comment (bendangelo) [inline src/map/utils/battleutils.cpp]: It binds for at least one second.

**Comment by bendangelo** [comment]
Thanks for your work. Adding this as a flag on the mob families table would work better.

**Comment by ymkc** [comment]
ok cool, np. just want to have it moving forward :)

@teschnei gonna take a look or want me to add it?

**Comment by xdemolish** [comment]
mob_family_mods, add a new MOBMOD and check that

**Comment by bendangelo** [comment]
A mob mod works as well

**Comment by teschnei** [comment]
I don't like using a mobmod for this. I thought about it earlier, but it
makes it kind of hard to remember when adding new families. Mob mods are
like.. Adding specific functionality to something, but uncharmable is a
pretty generic thing that should be in your face when adding a new family
(you should be actively presented with "is this family charmable?" Rather
than having the possibility of forgetting to check the family mobmods, if
that makes sense

**Comment by xdemolish** [comment]
ya
was under the impression its not really a significant number of families which are uncharmable hence the mobmod to avoid adding a 'charmable' member to every mobentity object (ya i realise you didnt say that but w/e ive already typed this far cantstopwontstop)

anyways, mobmod but set from the new field instead of mob_family_mods and check that in battleutils

**Comment by teschnei** [comment]
Yeah, I like that compromise. It keeps it overwritable by pool/script if
needed.

**Comment by Hozu** [comment]
Should the beastmen ecosystem check be part of the uncharmable flag set in the DB as well?

**Comment by Hozu** [comment]
Actually I just remembered that Nightmare mobs in dreamlands Dynamis are able to be charmed on DSP... might be a good idea to make the creation of the mobs for Dynamis/BCNMs/instances set the mobmod automatically, like how true aggro is currently enforced for Dynamis. That way it can be overridden if necessary.

**Comment by bendangelo** [comment]
Beast ecosystem should be moved to a flag in the family mob table. Good idea Hozu, dynamis mobs shouldn't be charmed and the flag should be set probably in mobutils.

**Comment by xdemolish** [comment]
nah it'll get set when adding the mob to the battlefield (when i eventually get round to finishing the rewrite)
though i guess master can have it set on gaining a confrontation effect

**Comment by ymkc** [comment]
> mobmod but set from the new field instead of mob_family_mods
add a new field to which table?

**Comment by xdemolish** [comment]
mob_family_system

**Comment by TeoTwawki** [comment]
This may have changed in retail since it was somewhat exploitable: back when I played placeholders for an NM were not able to be charmed, even though mobs of that family are normally able to be charmed. So the mobmod makes sense to me for script-ability for the reasons above and this, should we need to make PH's uncharmable in the future (not sure we do, but the ability will be there).

**Comment by xdemolish** [comment]
leaving for kj cause not sure about checking by ecosystem instead of just setting the flag and leaving it at that
also remember to leave a comment after updating a pull request so we know to have another look :)

**Comment by ymkc** [comment]
I added the ecosystem checks back to reduce the sql queries. I would also exclude battlefield mobs and flag them from the scripts (is there even any chamable battlefield mob apart from the furies?)

**Comment by teschnei** [comment]
there are giant queries that load mobs in zoneutils and instance_loader: just add the charmable column to the select list there and set it where every other parameter is set

**Comment by ymkc** [comment]
thanks for the input, works like a charm :)

code updated

**Comment by teschnei** [comment]
now remove m_charmable and set the mobmod in zoneutils/instanceutils

**Comment by ymkc** [comment]
could have seen that one ... done

**Pull request merged**